Feature: basic analytics visualizations

cap10morgan opened the pull request:
<img width="1379" height="997" alt="Screenshot 2025-10-03 at 8 49 18 AM" src="https://github.com/user-attachments/assets/601101f5-d075-4da9-83b0-42e54e867ede" />

Implements CORE-2767

I still need to add a few more metrics (once I confirm I can access them via `get_analytics`, but the basic structure is ready. This was meant as a research spike, so feel free to take it, leave it, or anything in between!

cap10morgan committed:
feat: Add basic monitoring component to status page

cap10morgan committed:
fix: Move type cast into type param

cap10morgan committed:
fix: Resolve pnpm-lock.yaml conflicts

dawsontoth commented on src/features/instance/operations/queries/getAnalytics.ts:
```diff
@@ -0,0 +1,46 @@
+				req.conditions = [{ attribute: 'path', value: metricConfig.path }];
+			}
+			const { data } = await instanceClient.post('/', req);
+			return data;
```
Can we type the response?

cap10morgan replied:
> Can we type the response?

Not easily. There's a _lot_ of variation, and with custom metrics it's pretty open-ended in the general case. I could put a very loose type on it that reflects that, though.

cap10morgan commented:
It's hard to see in the screenshot, but it also adds a new line for each node in a cluster. The new lines follow the Harper color palette (which may need some tweaking contrast-wise against this background color).

dawsontoth commented on src/features/instance/status/index.tsx:
```diff
@@ -3,5 +3,5 @@ import { CloudStatus } from '@/features/instance/status/CloudStatus';
 import { LocalStatus } from '@/features/instance/status/LocalStatus';
 
 export function StatusIndex() {
-	return isLocalStudio ? (<LocalStatus />) : (<CloudStatus />);
+	return isLocalStudio ? <LocalStatus /> : <CloudStatus />;
 }
```
We'll probably want to shift this to use a lazy loaded component so that the charting library only comes in when you visit the status page, not all the time. The profile page has an example of this.
```
const profileRoute = createRoute({
	getParentRoute: () => dashboardLayout,
	path: 'profile',
	component: lazyRouteComponent(async () => import('@/features/profile/index'), 'ProfileIndex'),
});

export const profileRoutes = [profileRoute];
```
You could also shift the common components you added to Local and Cloud Status into this index (essentially reverting them to what they were before, with all your changes up here instead).

cap10morgan committed:
fix: Add GetAnalyticsResponse type

Even this might be overly restrictive haha

dawsontoth commented:
Thoughts on ways we could handle the empty data state better? I'm guessing overlaying a bit of text to explain why it's empty would help. "No reads for this time period!"
<img width="2191" height="1460" alt="Screenshot 2025-10-03 at 11 32 41 AM" src="https://github.com/user-attachments/assets/9445eea9-f961-4ff7-91aa-9eddf1763522" />

dawsontoth commented on src/features/instance/status/components/Monitoring.tsx:
```diff
@@ -0,0 +1,108 @@
+					<Input
+						className="mr-2 w-16 inline-block"
+					  id="updateIntervalInput"
+					  type="text"
```
I recommend type="number" here. For fun, though, see what happens when you type `20a` and get NaN, particularly watch your network requests.

dawsontoth replied:
Alternatively, a dropdown would force the user to pick from some sane values instead. Less can go wrong then.

cap10morgan replied:
> Alternatively, a dropdown would force the user to pick from some sane values instead. Less can go wrong then.

Yeah, that's better. Done in https://github.com/HarperFast/hdbms/pull/786/commits/53e7f82c0f67e0858ef7e3d577a65b7485fe093e

dawsontoth commented on src/features/instance/operations/queries/getAnalytics.ts:
```diff
@@ -0,0 +1,52 @@
+			if (metricConfig.path) {
+				req.conditions = [{ attribute: 'path', value: metricConfig.path }];
+			}
+			const { data } = await instanceClient.post<GetAnalyticsResponse>('/', req);
```
Array?

cap10morgan replied:
> Array?

Haha yup!

cap10morgan commented:
> The units on the CPU graph are a bit perplexing to me. Is that showing user time? What does it cap at? If we can convert that to show a percentage, and force that graph to clamp to 0-100% (or whatever our max is) that would help with readability.

Yeah I'm not super clear on it either. This is where it gets calculated / set: https://github.com/HarperFast/harperdb/blob/007ccfe314624832be5ff322c212ac52ea5ff7ed/resources/analytics/profile.ts#L37-L98 I think ultimately we want it to be clear how it correlates to what we measure for billing purposes. Maybe @kriszyp has some more insight?

cap10morgan committed:
fix: Make GetAnalyticsResponse an array

cap10morgan committed:
fix: Remove old commented-out elements

cap10morgan committed:
fix: use type="number" for monitoring update interval

cap10morgan committed:
feat: Show explanation when no analytics data

cap10morgan committed:
chore: Move monitoring up to status/index & other cleanups

cap10morgan committed:
feat: Lazy load the status component

...so the charting libraries only get pulled in when / if users navigate there

cap10morgan committed:
feat: Switch update interval to drop down

This solves several data validation problems on the previous free-form number input

dawsontoth reviewed (approved):
I like it!
